src/logging.rs

```diff
@@ -0,0 +1,63 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use log::{Record, Metadata};
+use std::ffi::CString;
+use std::os::raw::c_char;
+use std::os::raw::c_int;
+
+pub type LogCallback = unsafe extern "C" fn(msg: *const c_char);
+
+struct LogProxy {level: log::Level, callback: LogCallback}
+
+impl log::Log for LogProxy {
+    fn enabled(&self, metadata: &Metadata) -> bool {
+        metadata.level() <= self.level
+    }
+
+    fn log(&self, record: &Record) {
+        if self.enabled(record.metadata()) {
+            let msg = format!("{} - {}", record.level(), record.args());
+            // println!("{}", msg);  // Print the message to stdout
+            let cb: LogCallback = self.callback;
+            let s = CString::new(msg).unwrap();

```

Comment:
even better: we could return the length of the string explicitly as well as the string (aka Pascal format), so that no conversion is needed here (it's pretty expensive to heap-allocate a new CString).

Reply:
The API uses `WGPURawString` (`const char *`) in other places where strings are communicated. Using a different format here would make the API less consistent. Do yo think it's worth it? (Your call. I just want to make sure we're on the same page :) )

Reply:
The problem is that heap-allocation here is too heavy if an application logs a log and still wants to run smoothly at run-time. I guess we can have that now and then improve without breaking the API. For example, we could have a thread-local `Vec<u8>` that we put the string contents in and append a null terminator before calling the user function. Again, we can follow up with that, it's fine to proceed with what you have now.